File: frontend/src/services/attendance-service.ts
import axios from 'axios';
import {GlobalConstants} from "../utils/GlobalConstants.ts";
import {Attendance} from "../models/Attendance.ts";
import {AttendanceStatus} from "../models/AttendanceStatus.ts";

export class AttendanceService {

    static baseUrl: string = GlobalConstants.baseUrl + 'attendance';

    static saveAttendance(username: string, lessonTimeId: number, status: AttendanceStatus): Promise<Attendance> {
        return axios.post(`${this.baseUrl}/save`, null, {
            params: {
                username,
                lessonTimeId,
                status
            }
        })
        .then(response => response.data)
        .catch(error => {
            console.error('Error saving attendance:', error);
            throw error;
        });
    }

    static getAttendanceByLessonAndDate(lessonId: number, date: string): Promise<Attendance[]> {
        return axios.get(`${this.baseUrl}/lesson`, {
            params: {
                lessonId,
                date
            }
        })
        .then(response => response.data)
        .catch(error => {
            console.error('Error fetching attendance:', error);
            throw error;
        });
    }
}

File: ubs/src/main/java/org/yigitcanyontem/ubs/service/AttendanceService.java
package org.yigitcanyontem.ubs.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.yigitcanyontem.ubs.domain.Attendance;
import org.yigitcanyontem.ubs.domain.Lesson;
import org.yigitcanyontem.ubs.domain.LessonTime;
import org.yigitcanyontem.ubs.domain.Users;
import org.yigitcanyontem.ubs.enums.AttendanceStatus;
import org.yigitcanyontem.ubs.repository.AttendanceRepository;
import org.yigitcanyontem.ubs.repository.LessonRepository;
import org.yigitcanyontem.ubs.repository.LessonTimeRepository;
import org.yigitcanyontem.ubs.repository.UsersRepository;

import java.util.Date;
import java.util.List;

@Service
public class AttendanceService {

    @Autowired
    private AttendanceRepository attendanceRepository;

    @Autowired
    private LessonRepository lessonRepository;

    @Autowired
    private LessonTimeRepository lessonTimeRepository;

    @Autowired
    private UsersRepository usersRepository;

    public List<Attendance> getAttendanceByLessonAndDate(Integer lessonId, Date date) {
        return lessonTimeRepository.findByLessonIdAndDate(lessonId, date).getAttendances();
    }

    public Attendance saveAttendance(String username, Integer lessonTimeId,AttendanceStatus status) {
        Users user = usersRepository.findByUsername(username)
                .orElseThrow(() -> new IllegalArgumentException("User not found with username: " + username));
        LessonTime lesson = lessonTimeRepository.findById(lessonTimeId)
                .orElseThrow(() -> new IllegalArgumentException("Lesson not found with id: " + lessonTimeId));

        Attendance attendance = Attendance.builder()
                .user(user)
                .lesson(lesson)
                .status(status)
                .recordedAt(new Date())
                .build();

        return attendanceRepository.saveAndFlush(attendance);
    }
}
